hack/module-graph.sh

```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env bash
+
+# Copyright 2020 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script generates a jpg image of our internal dependency graph.
+# It relies on go mod and dot to do so and should be run from K8s root.
+# TODO: Containerize the script to remove dependency issues with go mod and dot.
+
+error_exit()
+{
+	echo "$1" 1>&2
+	exit 1
+}
+
+echo "digraph module_dependencies {" > _output/module-dependencies.dot || error_exit "Failed to open DOT file"
+# Generating lines of the form " k8s_io_kubernetes -> k8s_io_api"
+# Excluding all non Kubernetes dependencies
+# Trimming away version info
+# Replacing non DOT (graph description language) characters with underscores 
+# Inserting needed arrow.
+# Indenting the line appropriately
+go mod graph | egrep "^.*k8s.io.*k8s.io.*$" | sed -E 's|@\S+ | |g' | sed -E 's|@\S+$||g' | sed -E 's/[\.\/\-]/_/g' | sed -E 's| | -> |g' | sed -E 's|^|    |g' >> _output/module-dependencies.dot || error_exit "Failed to generate dependencies in DOT file"
```

Comment:
since we have some k8s.io/... modules that aren't in staging, something like this would let you limit the graph to just the in-tree modules:

```
relevant_modules="k8s.io/(kubernetes|$(ls staging/src/k8s.io | tr '\n' '|' | sed 's#|$##' ))"
go mod graph | egrep "^${relevant_modules}(@.*| )${relevant_modules}@.*$" | ...
```

Reply:
with that change:
![module-dependencies](https://user-images.githubusercontent.com/980082/85618891-745a1400-b62f-11ea-890c-145fb053d7df.jpg)

Reply:
without that change:
![module-dependencies](https://user-images.githubusercontent.com/980082/85618977-95bb0000-b62f-11ea-9595-814c853c4e4c.jpg)

Reply:
Little torn. The first graph is definitely simpler, easier to read and nicer looking. Most of the nodes which are removed are leaf nodes and I don't think they are really providing any value. However at least three of the nodes which the change removes are not leaf nodes (utils, kube_openapi, apiserver_network_proxy_konnectivity_client). Tempted to put in flag to let user decide on which. Also found we have duplicate lines which is part of why the first is so messy. Added fix to dedupe lines.

Reply:
![module-dependencies](https://user-images.githubusercontent.com/947971/85754115-ad42c900-b6c1-11ea-985d-a1e5a3aca4d0.jpg)

Reply:
I don't feel that strongly, I just thought the goal was to find dependencies between modules that live in k8s.io/kubernetes

Reply:
I was thinking more understanding the dependencies between modules we (kubernetes) own/control.